### PR TITLE
Add documentation content for Sphinx/RTD

### PR DESCRIPTION
**Pull request by willingc**
Addresses #101

### CODE REVIEW

**Comment by willingc** [comment]
@parente Here's the first version of the Sphinx documentation. Once it is merged, I will circle back and update the README.md file and other files that depend on the links of the deployed RTD version.

In general, I moved most of the README content into the docs/source markdown files.

File: README.md [hunk 1]
```diff
@@ -1,334 +1,27 @@
...
-```
-cd kernel_gateway
-make dev
-```
-
-To access the gateway instance:
-
-1. Run `docker-machine ls` and note the IP of the dev machine.
-2. Visit http://THAT_IP:8888/api in your browser. (Note that the route `/api/kernels` is not enabled by default for security. See the `--KernelGatewayApp.list_kernels` parameter above.)
+### Setting up a development system
```
Comment (parente) [inline README.md]: Can we retain the Develop section which has the instructions for setting up a dockerized dev environment? We can certainly have other options, but I think having something here to start while alternative approaches are written is better than nothing.

File: docs/source/changelog.md [hunk 2]
```diff
@@ -1,8 +1,56 @@
-# Summary of changes in Jupyter Kernel Gateway
+# Summary of changes
 
-See `git log` for a more detailed summary.
+See `git log` and `CHANGELOG.md` for a more detailed summary of changes.
```
Comment (parente) [inline docs/source/changelog.md]: Are we going to have to copy/paste the CHANGELOG into the docs now before every release? Would be nice if the CHANGELOG automatically got pulled into the sphinx build. (Little bit of make magic?)

File: docs/source/config-options.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+KernelGatewayApp options
+------------------------
+--KernelGatewayApp.allow_credentials=<Unicode>
+    Default: ''
+    Sets the Access-Control-Allow-Credentials header. (KG_ALLOW_CREDENTIALS env
+    var)
+--KernelGatewayApp.allow_headers=<Unicode>
+    Default: ''
+    Sets the Access-Control-Allow-Headers header. (KG_ALLOW_HEADERS env var)
+--KernelGatewayApp.allow_methods=<Unicode>
```
Comment (parente) [inline docs/source/config-options.md]: It'd be really nice to get this to autogenerate on sphinx build as well. Food for thought for later.

File: README.md [hunk 4]
```diff
@@ -1,334 +1,27 @@
...
+* A Jupyter Notebook server-compatible HTTP API used for requesting kernels
+  and talking the [Jupyter kernel protocol](http://jupyter-client.readthedocs.org/en/latest/messaging.html)
+  with the kernels over Websockets
+* A HTTP API defined by annotated notebook cells that maps HTTP verbs and
+  resources to code to execute on a kernel
 
-The server launches kernels in its local process/filesystem space. It can be containerized and scaled out by a cluster manager (e.g., [tmpnb](https://github.com/jupyter/tmpnb)).
+The server launches kernels in the server's local process/filesystem space.
+The server can be containerized and scaled out by a cluster manager (e.g.,
+[tmpnb](https://github.com/jupyter/tmpnb)).
```
Comment (parente) [inline README.md]: When the doc is posted, we should add a direct link to it here "See the features page in the documentation for more details."

File: docs/source/uses.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  running near big data (e.g., interactive gateway to Spark)
+* Enable a new breed of non-notebook web clients to provision and use 
+  kernels (e.g., dashboards using 
+  [jupyter-js-services](https://github.com/jupyter/jupyter-js-services))
+* Scale kernels independently from clients (e.g., via 
+  [tmpnb](https://github.com/jupyter/tmpnb), [Binder](https://mybinder.org),
+  or your favorite cluster manager)
+* Create microservices from notebooks via 
+  [`notebook-http` mode](#notebook-http-mode)
+
```
Comment (parente) [inline docs/source/uses.md]: I'm a little worried about burying these in the doc. We meant them as aids for understanding what the heck you'd want to do with the KG after reading the very technical introduction which assumes a lot of knowledge about the underpinnings of Jupyter.

**Comment by parente** [comment]
Thanks @willingc. I left a few questions/comments.

Once this goes in, what do we need to do to get the RTD site rebuilding? And when does it need to rebuild? I'll need a bit of a tutorial on the maintenance flow that the other projects are following. For example, are they tagging and building version specific copies of the doc to RTD?

File: README.md (hunk 1 above)
Comment (willingc) [inline README.md]: Absolutely. I may edit it down for the README. Will push a change to this PR later today.

File: docs/source/changelog.md (hunk 2 above)
Comment (willingc) [inline docs/source/changelog.md]: The RTD version of the changelog.md will likely evolve over time to summary-changes.md which will likely have a more concise description of changes.

File: docs/source/config-options.md (hunk 3 above)
Comment (willingc) [inline docs/source/config-options.md]: Sure. We do that on other repos. For now let's leave as is and if time permits I will migrate to the autogenerate.

File: README.md (hunk 4 above)
Comment (willingc) [inline README.md]: That's the plan :D

File: docs/source/uses.md (hunk 5 above)
Comment (willingc) [inline docs/source/uses.md]: I think it's good to keep these in the docs. I will update the README to share the real world uses of KG as well as a link to this section of the docs.

**Comment by willingc** [comment]
@parente Thanks for the thoughtful review.

Here are two docs that will help with understanding the details of having the docs build on RTD:
[General RTD workflow for Jupyter](https://github.com/willingc/calpoly-cnw/blob/master/doc-workflow/docs-readthedocs.md)
[Enable RTD site building on commits](https://github.com/willingc/calpoly-cnw/blob/master/doc-workflow/docs-rtd-hook.md)

I'm happy to set up RTD and the hook since I've done it a bunch now.

As for maintenance flow of docs, I'm still working on the overall process. Here's the ideal workflow that I'm working toward for all the Jupyter projects:
- When a release is readied for alpha or beta, update the docs to reflect changes/features/etc.
- Prior to tagging a release in GitHub, it would be ideal for the project leads to make sure that the docs are up to date. (Note: RTD can only build docs for either a tagged commit or HEAD of master. If the docs are not included in the tagged release, RTD will incorrectly report the version number of the docs.) In short, the tag of a release applies to the technical content and docs as of the release date.

This project is looking great. Nicely done @parente.

File: docs/source/uses.md (hunk 5 above)
Comment (parente) [inline docs/source/uses.md]: As long as the readme has some text to hook the reader interest and then links to use cases / this page in docs, that's fine.

**Comment by parente** [comment]
Thanks for the responses. I'm only hesitating on merging this because the README is barren now and the doc site is not up yet. Is it possible to get the site up and running and the README repopulated with the key information I noted above before merging? Or do we really have to merge first?

**Comment by willingc** [comment]
@parente Here's an update with your suggested changes. I made the README more comprehensive. We can merge if you would like. After it's merged I can correct any needed live links :D

**Comment by willingc** [comment]
Buttons are challenging...

**Comment by parente** [comment]
LGTM.

FWIW, I'm nearly done with a draft blog post about KG for @Carreau. I think we should hold off on posting that until the docs are up and most links are working  so that people who click through get an OK first impression.

**Comment by parente** [comment]
> I'm happy to set up RTD and the hook since I've done it a bunch now.

I set the webhook. Not sure what I need to do to bootstrap it first time.

**Comment by parente** [comment]
And now it's magically there: http://jupyter-kernel-gateway.readthedocs.org/en/latest/

+(:100: x :100:) @willingc for getting this ball rolling!

**Comment by willingc** [comment]
Yeah, I set it up a few minutes ago on RTD. I'm going to do a bit of clean up on them. But we should be in good shape by tonight.

**Pull request merged**